src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/ArrayMarshaller.cs

```diff
@@ -107,7 +108,12 @@ public ArrayMarshaller(T[]? array, Span<byte> buffer, int sizeOfNativeElement)
         /// </remarks>
         public ReadOnlySpan<byte> GetNativeValuesSource(int length)
         {
-            return _allocatedMemory == IntPtr.Zero ? default : _span = new Span<byte>((void*)_allocatedMemory, length * _sizeOfNativeElement);
+            if (_allocatedMemory == IntPtr.Zero)
+                return default;
```

Comment:
Is _allocatedMemory == IntPtr.Zero a common case?  I'm wondering what would go wrong if you just deleted this if block.  Presumably then the only way this would fail is if this method could be called with a length in appropriate for whatever _allocatedMemory is?

Reply:
This is an attempt to handle cases where the returned values are in conflict with the allocation. I don't think we can remove this.

Reply:
> This is an attempt to handle cases where the returned values are in conflict with the allocation.

So I can better understand, can you give an example of what this means?

Reply:
Consider the following case:

```csharp
[LibraryImport(NativeExportsNE_Binary, EntryPoint = "get_long_bytes")]
[return:MarshalAs(UnmanagedType.LPArray, SizeConst = sizeof(long))]
public static partial byte[] GetLongBytes(long l);
```

This states the size of the returned array is 8. However, this doesn't encode the case where the caller returns `null`. That means we may have a `null` returned but a well-defined length and with that would internally create a `Span<byte>(null, 8)`. Checking if the allocation is `IntPtr.Zero` (i.e., `null`) means the callee can return `null` even in cases where constant length is defined and the marshaller will fallback to creating a "default" view of memory instead of creating a `Span<T>` that is unsupported - `null` with a non-zero length.